build-tools/packages/build-cli/src/library/markdown.ts

```diff
@@ -41,3 +42,105 @@ export function remarkHeadingLinks(): (tree: Node) => void {
+ * Additionally, it captures any whitespace characters that follow the matched pattern.
+ */
+const ADMONITION_REGEX = /(\[!(?:CAUTION|IMPORTANT|NOTE|TIP|WARNING)])(?!\s*$)(\s*)/gm;
+
+/**
+ * A regular expression to remove single line breaks from text. This is used to remove extraneous line breaks in text
+ * nodes in markdown. This is useful because GitHub sometimes renders single line breaks, and sometimes it ignores them
+ * like the CommonMark spec describes. Removing them ensures that markdown renders as expected across GitHub.
+ */
+const SOFT_BREAK_REGEX = /$[^$]/gms;
+
+/**
+ * A remarkjs/unist plugin that strips soft line breaks. This is a workaround for GitHub's inconsistent markdown
+ * rendering in GitHub Releases. According to CommonMark, Markdown paragraphs are denoted by two line breaks, and single
+ * line breaks should be ignored. But in GitHub releases, single line breaks are rendered. This plugin removes the soft
+ * line breaks so that the markdown is correctly rendered.
+ */
+export function stripSoftBreaks(): (tree: Node) => void {
+	return (tree: Node): void => {
+		// strip soft breaks
+		visit(tree, "text", (node: { value: string }) => {
+			node.value = node.value.replace(SOFT_BREAK_REGEX, " ");
+		});
+
+		// preserve GitHub admonitions; without this the line breaks in the alert are lost and it doesn't render correctly.
+		visit(tree, "blockquote", (node: Node) => {
+			visit(node, "text", (innerNode: { value: string }) => {
+				// If the text is an admonition title, split
+				innerNode.value = innerNode.value.replace(ADMONITION_REGEX, "$1\n");
+			});
+		});
+	};
+}
+
+export function removeSectionContent(options: { heading: string | RegExp }): (
+	tree: Root,
+) => void {
+	return function (tree: Root) {
+		headingRange(tree, options.heading, (start, nodes, end, info) => {
+			console.log(`removing section ${options.heading}`);
+			return [
+				start,
+				// No child nodes - effectively empties the section.
+				end,
+			];
+		});
+	};
+}
+
+export function removeHeadingsAtLevel(options: { level: 1 | 2 | 3 | 4 | 5 | 6 }): (
+	tree: Root,
+) => void {
+	return (tree: Root) => {
+		let headingRemoved = false;
+
+		visit(
+			tree,
+			"heading",
+			(node: Heading, index: number | undefined, parent: Parent | undefined) => {
+				if (
+					!headingRemoved &&
+					// node.children?.[0].type === "text" &&
+					// node.children[0].value === "⬆️ Table of contents" &&
+					node.depth === options.level &&
+					index !== undefined
+				) {
+					parent?.children.splice(index, 1);
+					headingRemoved = true;
+					return [SKIP, index];
+				}
+			},
+		);
+	};
+}
+
+export function updateTocLinks(options: { newUrl: string }): (tree: Root) => void {
+	const { newUrl } = options;
+
+	return (tree: Root) => {
+		visit(tree, "link", (node: Link) => {
+			if (
+				node.children?.[0].type === "text" &&
+				node.children[0].value === "⬆️ Table of contents"
```

Comment:
Can we give these link nodes a class or something similar to identify them? If we ever change the text this seems prone to breaking (we'd have to update the generation logic and then this bit too, right?).

Reply:
I think such a thing is possible but not without writing some custom mdast nodes or something like that. I don't think it's worth it right now. But I can make this string a shared constant.